File: 연결요소의개수.py
from collections import deque
n,m = map(int,input().split())
check= [False]*(n+1)
g=[[] for _ in range(n+1)]

for _ in range(m):
  u,v = map(int,input().split())
  g[u].append(v) 
  g[v].append(u)

def bfs(start):
  global check
  q=deque()
  check[start] =True
  q.append(start)
  while q: 
    x= q.popleft()
    for y in g[x]:
      if not check[y]:
        check[y] =True
        q.append(y)
ans = 0
for i in range(1,n+1):
  if not check[i] :
    ans+=1
    bfs(i)
print(ans)

File: 연결요소의개수.cpp
#include <iostream>
#include <queue>
#include <algorithm>
#include <vector>
using namespace std;

const int MAX = 1001;
vector<int> g[MAX];
bool check[MAX];

void dfs(int x){
  check[x] = true;
  for(int i=0 ; i <g[x].size(); i++){
    int y = g[x][i];
    if(!check[y]) dfs(y);
  }
}

void bfs(int start){
  queue<int> q;
  check[start]=true; q.push(start);
  while(!q.empty()){
    int x = q.front(); q.pop();
    for(int i=0; i<g[x].size(); i++){
      int y = g[x][i];
      if(!check[y]){
        check[y]=true; q.push(y);
      }
    }
  }
}

int main(){
  int n,m;
  cin>>n>>m;
  for(int i=0; i<m; i++){
    int u,v;
    cin>>u>>v;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  int ans =0;
  for(int i=1; i<=n; i++){
    if(!check[i]){
      ans++;
      // dfs(i);
      bfs(i);
    }
  }
  cout<<ans<<'\n';
}